### DISCUSSION TITLE
Nodal loads

### DISCUSSION THREAD

**Discussion by sidharthsarmah** [idaholab/moose#26768] [Q&A Modules: Solid mechanics] [answered]
Hi @GiudGiud ,

I applied a nodal loads at one of the nodes for a bending test, but the higher loads resulted in an infinite stress at the notes.

Could you suggest if I should 
1. Use a pressure BC over some nodes OR
2. Distribute these nodal loads over some nodes using `UserForcingFunctionNodalKernel`.

<img width="531" alt="image" src="https://github.com/idaholab/moose/assets/110927986/009362ae-fbb7-4e03-9d81-7d4aceebd52e">

**Comment by GiudGiud** [comment]
Hello

I dont know. I reckon you should try both since these are input files modifications and should not take too long. 

Guillaume

**Reply by sidharthsarmah** [reply]
Thanks, @GiudGiud , but there is a discrepancy when I applied the same concentrated loads at nodes in the same problem in ABAQUS, there is no deformation. Is there something I am missing while considering the `NodalKernal`?
![image](https://github.com/idaholab/moose/assets/110927986/345feb18-d8c3-4dec-adaf-527ba7a10f8a)

**Reply by GiudGiud** [reply]
maybe the magnitude is different?
what do you have in Abaqus and what are you putting in MOOSE?

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , Its the same magnitude.

In MOOSE:

**Reply by GiudGiud** [reply]
same mesh too?

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud,

I tried for linear elastic case without CZM where the loading works and continues till completion, which means loading is okay.
Just for a comparison, the VM stress for elastic case goes in MOOSE goes to 94000 MPa without nodal failure, while in the plastic case in ABAQUS, it goes to 2583 MPA, which smaller than elastic and should be the same for MOOSE.
I think maybe it is with the stress-strain curve implementation. Could you please check, if this is correct way to incorporate an stress-strain curve in MOOSE?

**Reply by GiudGiud** [reply]
why are your stress units in MPa?
The mesh is in meters right?

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , no its in `mm` and stress is in `MPa (N/mm^2)`, loads in `N`

**Reply by GiudGiud** [reply]
the function is used like that:
```
    const Real strain_old = this->_effective_inelastic_strain_old[_qp];
    return _hardening_function->value(strain_old + scalar) - _yield_stress;
```
so is the 1st column in the same units as the strain?
and has the yield stress NOT been substrated already from the 2nd column

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , yes the first column is the plastic strain, and second is the flow stress, which has yet not been subtracted.

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , do you suggest I tabulate with something like this:
where 1500 is the yield stress

**Reply by GiudGiud** [reply]
no the yield stress is substracted already so that should be fine. I dont see what is wrong.

Someone from the solid mechanics community will have to help

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud thanks, I think then the error is something else, which I shall try to work on.

The only thing I am confuse is it worked for elastic case but when I insert plasticity there is compilation error at higher loads.

<img width="612" alt="image" src="https://github.com/idaholab/moose/assets/110927986/bb6b7e62-59b6-47f4-a401-2f6f4862f333">

**Reply by GiudGiud** [reply]
So plasticity without the bottom loads is fine, but plasticity with those does not converge?

**Comment by bwspenc** [comment]
Reading this discussion, it sounds like you're using CZM, is that right? Do you have CZMs running on the interfaces right where you're applying the load?

I would suggest backing off some of the complexity of the model to sort this out (e.g., turning off some of the things like CZMs and plastic material, and just running this with elastic material).

**Reply by bwspenc** [reply]
Actually, now that I've gone back in the thread, it sounds like maybe you successfully did just that.

**Reply by bwspenc** [reply]
I'm really having a tough time following what the issue is from this thread. If you are able to share an input file I can take a look.

**Reply by sidharthsarmah** [reply]
Hi @bwspenc @GiudGiud , I tried elastic case where the force worked, but plastic case don't work.

**Reply by bwspenc** [reply]
Thanks! Could you share the mesh too?

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , @bwspenc
This is something I found:
When I changed the BC from Load to displacement type keeping everything similar (plastic case) I get a full deformation without breakage of mesh unlike in loads.
<img width="881" alt="image" src="https://github.com/idaholab/moose/assets/110927986/3073ac6c-2f10-42f3-8fc1-2132a803be74">

**Reply by bwspenc** [reply]
Is it possible that you were just applying too high of a load with the NodalKernel? I typically use Dirichlet BCs for this sort of thing whenever possible it generally works better in any code.

**Reply by sidharthsarmah** [reply]
Hi @bwspenc @GiudGiud , thanks but the problem is I am validating this model that has a loading BC, could you suggest if there is any way because when I have used `Dirichlet BCs`, the load in the the same nodes goes to higher values (~900), but when. apply a `NodalKernel`, the same load does not reach that values, its very confusing, please suggest?

**Reply by bwspenc** [reply]
I believe that you shared the mesh and one of the CSV files earlier today, but I don't see those now. Maybe I'm just missing something! If you can send all the files I need to run this, I can take a look.

**Reply by bwspenc** [reply]
I tried running your model with both force and displacement control. I added the following postprocessors to look at the average displacements at those two nodes as well as the reaction forces at them individually:

```
[Postprocessors]
  [top_disp]
    type = NodalL2Norm
    boundary = 'top_left top_right'
    variable = disp_y
  []
  [top_left_react]
    type = NodalL2Norm
    boundary = 'top_left'
    variable = force_y
  []
  [top_right_react]
    type = NodalL2Norm
    boundary = 'top_right'
    variable = force_y
  []
[]
```
It looks to me like the model is behaving similarly in both cases. You're using pretty large time steps in the model you sent, so after just a few time steps it's getting up to the force level of 1462.5 (90*16.25) where your function plateaus out for a while, and you're getting pretty significant nonlinearity at that point. It seems like it would be expected behavior to end up with yielding right at the supports under those conditions.

**Reply by sidharthsarmah** [reply]
Hi @bwspenc , thanks.

You mean at say at 90 x 16.25N (around 3600s) the model breaks the mesh because of large loads. In actuality, the final load goes to 468000 s and applies 940 x16.25N load and is not occur as huge plasticity is involved beyond realistic values.

Can you please check once how it behaves at 468000s?

**Reply by bwspenc** [reply, marked as answer]
I stopped the model long before that time because it was really struggling to converge, which seems to be because the loads applied were so high.

One thing occurred to me about your comparison with another code: I'm sure you already are aware of this, but we don't have a way to specify the out-of-plane thickness for 2D models. The thickness of the model is 1, so applied loads would need to be divided by the thickness and the reactions from the model would need to be multiplied by the thickness. Are you positive you have that right?

**Reply by sidharthsarmah** [reply]
Hi @bwspenc ,

1. You mean to say for 2D cases in MOOSE, the thickness is by default considered as 1mm (if I have other dimensions in mm)?
2. The loads which I obtained at the nodes are (Load x 1mm) in N.mm.
3. As an example. when I say the load in the input file is 900, does it mean 900N or 900N/mm?
4. But does not dividing by length in the input and multiplying by a length at the output, will result in the same unit, I am a bit confused about that.

**Reply by bwspenc** [reply]
1) Yes, the thickness is 1 mm if your other dimensions are mm
2) The loads coming out from the model would be in N/mm. You would multiply them by your part's thickness in mm to get the loads in N.
3) If you are specifying a load of 900, that would be applied to your model as 900 N/mm. 
4) I keep things straight by thinking of the model as representing a unit thickness of the component.

**Reply by sidharthsarmah** [reply]
Thanks @bwspenc

**Reply by bwspenc** [reply]
I'll add that we really should add the capability for the user to specify the out-of-plane thickness. It's something that hasn't been a high priority because you can work around it, but it would help make the code easier to use!

**Reply by bwspenc** [reply]
@sidharthsarmah was your issue related to handling of the thickness, then? I just wanted to make sure you got it working.

**Reply by sidharthsarmah** [reply]
Hi @bwspenc @GiudGiud , yes, I was unaware of it the whole time and its finally sorted, thank you so much)

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , I am working on a 3 point bending problem. I feel the bottom BCs should be a roller supports. But as I put:
```
  [./fix_y]
    type = ADDirichletBC
    variable = disp_y
    value = 0.0
    boundary = 'bottom'
  [../]

  [./load_y]
    type = ADFunctionDirichletBC
    variable = disp_y
    function = '0.002*t'
    boundary = 'top'
  [../]
```
![image](https://github.com/idaholab/moose/assets/110927986/618eb141-591c-49e6-9fe9-e339f0769171)


There is convergence issue. But when I add:
```
  [./fix_y]
    type = ADDirichletBC
    variable = disp_x
    value = 0.0
    boundary = 'bottom'
  [../]
```

It works as a fixed joint which is not actually the case and there is extensive deformation at the notch and bottom, please suggest, how to get rid of the broken mesh.

<img width="725" alt="image" src="https://github.com/idaholab/moose/assets/110927986/3b0089b1-eef7-4d96-9ced-024c59813ccc">

**Reply by GiudGiud** [reply]
isnt the mesh a little coarse at the top where the load is applied?
and same question near the support on the bottom

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud ,
Yes, my finding is using a finer mesh for disp BC results in no convergence.
Using finer mesh works for loading BC where the load gets distributed to other nodes, but during displacement BC having a lot of nodes result in no displacement of the particular node in which the displacement BC is applied.

Please correct if this makes sense?

**Reply by GiudGiud** [reply]
Did you set `preset = false` to those dirichletBCs for displacement?
The problem with a fine mesh is that a large preset displacement ruins the elements, makes it inverted and this causes convergence issues
